### PR TITLE
Backport C++17 type_traits and cuda::std::byte to C++14

### PR DESCRIPTION
**Pull request by jrhemstad**
Broken off from https://github.com/NVIDIA/libcudacxx/pull/10 

- [x] Backports `<type_traits>` features from C++17 to make them available in C++14

- [x] Backports tests of type traits to make them supported in C++14

- [x] Backports `cuda::std::byte` to be available in C++14 (this was necessary to make some tests pass as a result of the `<type_traits>` changes

### CODE REVIEW

**Comment by jrhemstad** [comment]
@griwes are there `<type_traits>` tests I'm missing that should also be back ported?

**Review by griwes — APPROVED** [review]
Looking at the diff made me realize that you're also enabling the variable templates for the traits, but not the tests around the variable templates, like here: https://github.com/NVIDIA/libcudacxx/blob/main/.upstream-tests/test/std/utilities/meta/meta.unary/meta.unary.prop/is_abstract.pass.cpp#L24-L29. I'll leave it up to you whether you want to enable those in this PR or if you want to do that in a future one to help you folks move to this repo in RAPIDS, since those tests will quite obviously pass.

File: libcxx/include/type_traits [hunk 1]
```diff
@@ -4843,7 +4843,7 @@ using _IsCharLikeType = _And<is_standard_layout<_CharT>, is_trivial<_CharT> >;
 
 _LIBCUDACXX_END_NAMESPACE_STD
 
-#if _LIBCUDACXX_STD_VER > 14
+#if _LIBCUDACXX_STD_VER > 11
```
Comment (griwes) [inline libcxx/include/type_traits:4846]: Does the corresponding endif need to be updated here?

Comment (jrhemstad) [reply libcxx/include/type_traits:4846]: Done.

**Comment by jrhemstad** [comment]
> Looking at the diff made me realize that you're also enabling the variable templates for the traits, but not the tests around the variable templates, like here: https://github.com/NVIDIA/libcudacxx/blob/main/.upstream-tests/test/std/utilities/meta/meta.unary/meta.unary.prop/is_abstract.pass.cpp#L24-L29.

Thanks for noticing that. I didn't see the guards in the tests themselves. I updated them all to test in C++14 and they all passed.

**Comment by griwes** [comment]
Kicked off internal CI at CL 29255931.

**Comment by griwes** [comment]
It looks like MSVC doesn't accept `__is_aggregate` in its C++14 mode. You'll need to make its definition go away there. (It seems to me that guarding the [definition of `_LIBCUDACXX_IS_AGGREGATE` in libcxx/include/__config](https://github.com/NVIDIA/libcudacxx/blob/main/libcxx/include/__config#L866) with a language version check should be enough, but I have no Windows dev environment set up to confirm.)

**Comment by jrhemstad** [comment]
> It looks like MSVC doesn't accept `__is_aggregate` in its C++14 mode. You'll need to make its definition go away there. (It seems to me that guarding the [definition of `_LIBCUDACXX_IS_AGGREGATE` in libcxx/include/__config](https://github.com/NVIDIA/libcudacxx/blob/main/libcxx/include/__config#L866) with a language version check should be enough, but I have no Windows dev environment set up to confirm.)

I added a MSVC version check for >19.14.

**Comment by griwes** [comment]
Re-running internal CI, CL 29261259.

**Comment by griwes** [comment]
`std/language.support/support.types/byte.pass.cpp` passes unexpectedly on GCC in C++14 mode which causes a fail. Other than that it looks like we are good to go. Can you fix its keywords and remove the merge commit that's still in this PR? I'll merge it then.

**Comment by griwes** [comment]
Clarification: it's the _libc++_ version of the test that's XPASSing, in `libcxx/test`, not the libcu++ one.

**Pull request closed without merging**